src/lib/support/PersistentArray.h

```diff
@@ -0,0 +1,317 @@
+     * @param value [in, out] Used as an input for comparison, and output for the rest of the values.
+     * @param do_create When true, if no matching entry is found, the entire value will be stored
+     *                  as a new entry in the table.
+     * @return CHIP_NO_ERROR on success,
+     *         CHIP_ERROR_NOT_FOUND if no entry matches the given input.
+     */
+    CHIP_ERROR Get(EntryType & value, bool do_create = false)
+    {
+        CHIP_ERROR err = this->Load(true);
+        VerifyOrReturnError(CHIP_ERROR_NOT_FOUND == err || CHIP_NO_ERROR == err, err);
+
+        // Check existing
+        for (size_t i = 0; i < this->mCount; ++i)
+        {
+            EntryType & e = mEntries[i];
+            if (e.Compare(value))
+            {
+                ReturnErrorOnFailure(value.Copy(e));
+                return CHIP_NO_ERROR;
+            }
+        }
+        // Insert last ?
+        VerifyOrReturnError(do_create, CHIP_ERROR_NOT_FOUND);
+        VerifyOrReturnError(this->mCount < this->mLimit, CHIP_ERROR_INVALID_LIST_LENGTH);
+        ReturnErrorOnFailure(mEntries[this->mCount++].Copy(value));
+        return this->Save();
+    }
+
+    /**
+     * @brief Removes the stored entry at a given index. Remaining entries are
+     *        shifted one position down.
+     * @param index Zero-based position within the table.
+     * @return CHIP_NO_ERROR on success,
+     *         CHIP_ERROR_NOT_FOUND if index is greater than the index of the last entry on the table.
+     */
+    CHIP_ERROR Remove(size_t index)
+    {
+        ReturnErrorOnFailure(this->Load(true));
+        VerifyOrReturnError(index < this->mLimit, CHIP_ERROR_NOT_FOUND);
+        VerifyOrReturnError(index < this->mCount, CHIP_ERROR_NOT_FOUND);
+        EntryType old;
+        old.Copy(mEntries[index]);
+        this->mCount--;
+        for (size_t i = index; i < this->mCount; ++i)
+        {
+            mEntries[i].Copy(mEntries[i + 1]);
+        }
+        ReturnErrorOnFailure(this->Save());
+        OnEntryRemoved(old);
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * @brief Removes the value matching the given entry. The Compare() method
+     *        is used to identify the matching entry within the table.
+     * @param entry Identifies the value to be removed.
+     * @return CHIP_NO_ERROR on success,
+     *         CHIP_ERROR_NOT_FOUND if no entry matches the given input.
+     */
+    CHIP_ERROR Remove(const EntryType & value)
+    {
+        ReturnErrorOnFailure(this->Load(true));
+        for (size_t i = 0; i < this->mCount; ++i)
+        {
+            if (mEntries[i].Compare(value))
+            {
+                return Remove(i);
+            }
+        }
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * @brief Removes all the entries in the table.
+     * @return CHIP_NO_ERROR on success.
+     */
+    CHIP_ERROR RemoveAll()
+    {
+        ReturnErrorOnFailure(this->Load());
+        for (uint16_t i = 0; i < mLimit; ++i)
```

Comment:
Why is this going to mLimit (which includes un-initialized data, right?) instead of just mCount?